File: SeryeongLee/week3/14501_퇴사.py
# https://deeply-silence-9a4.notion.site/81e353a993164d1287853ca41cd8a12d?pvs=4

import sys 
input = sys.stdin.readline
T = []
P = []
N = int(input())

for i in range(N):
    arr = list(map(int, input().split()))
    T.append(arr[0])
    P.append(arr[1])
ans = 0

def solution(day, sum):
    global ans
    if day == N: # 현재 날짜 
        ans = max(ans, sum) # 저장된 ans와 
        return
    if day > N: # N날을 넘으면 벗어나기 
        return
    solution(day + T[day], sum + P[day]) # 현재작업 기준으로 현재날짜+ T[day] 현재이익+ P[day]
    solution(day + 1, sum) # 현재 작업을 건너뛰고 다음날짜 + 현재 총 이익 
    
solution(0,0) # 초기 시작
print(ans)

File: SeunghyunSon/week3/14501_퇴사.cpp
#include <iostream>
#include <algorithm>

using namespace std;

int dp[16] = { 0, };
int cost[16] = { 0, };
int day[16] = { 0, };

void solution(int n) {
	for (int i = n; i > 0; i--) {
		if (i + day[i] > n + 1){
			dp[i] = dp[i + 1];
		} else {
		    // 마지막 상담 종료 날짜가 퇴사 날짜를 넘어가는 것을 방지하기 위해
		    // dp[i+day[i]]와 비교해야 함
			dp[i] = max(dp[i + 1], cost[i] + dp[i + day[i]]);
		}
	}
}

int main(void) {
    
	int n;
	cin >> n;

	for (int i = 1; i <= n; i++) {
		cin >> day[i] >> cost[i];
	}

	solution(n);

	cout << dp[1] << '\n';
	
	return 0;
}